src/coreclr/jit/emitxarch.cpp

```diff
@@ -2314,15 +2314,38 @@ bool emitter::HasMaskReg(const instrDesc* id) const
     return false;
 }
 
+//------------------------------------------------------------------------
+// AbsRegNumber: Returns the register value to be used for encoding.
+// JIT internally represents registers using regnumber 0 - (REG_STK-1).
+// For encoding, we need to separate out GPR, SIMD and K registers.
+//
+// Arguments:
+// reg -- The register being encoded.
+//
+// Return Value:
+// The register value to be used for encoding.
+regNumber AbsRegNumber(regNumber reg)
+{
+    assert(reg < REG_STK);
+    if (reg >= KBASE)
+    {
+        return (regNumber)(reg - KBASE);
+    }
+    else if (reg >= XMMBASE)
+    {
+        return (regNumber)(reg - XMMBASE);
+    }
+    return reg;
+}
+
 // Returns true if using this register will require a REX.* prefix.
 // Since XMM registers overlap with YMM registers, this routine
 // can also be used to know whether a YMM register if the
 // instruction in question is AVX.
 bool IsExtendedReg(regNumber reg)
 {
 #ifdef TARGET_AMD64
-    // TODO-XArch-apx: extend the gpr test, extended gprs should be from r8 to r31 after apx.
-    return ((reg >= REG_R8) && (reg <= REG_R15)) || ((reg >= REG_XMM8) && (reg <= REG_XMM31));
+    return ((reg >= REG_R8) && (reg <= REG_R23)) || ((reg >= REG_XMM8) && (reg <= REG_XMM31));
```

Comment:
likewise here, I am not sure if this is correct. For non-APX, extended reg should be from `R16~R23`.

Reply:
Currently existing(before this PR) `isHighGPReg()`  is a dummy method Ruihan added which is not used. `isHighGPReg()` should return `true` only when reg allocator has chosen a eGPR. It essentially tells the emitter that we need to chose an encoding that supports eGPR(`REX2` for example) and needs to set the corresponding bits (`REX2.X4` for example)

`IsExtendedReg()` needs to return true if we have to use REX or REX2. This means `true` for anything higher that `R8`